src/main/java/uk/gov/hmcts/divorce/noticeofchange/event/SystemApplyNoticeOfChange.java

```diff
@@ -0,0 +1,125 @@
+import uk.gov.hmcts.divorce.noticeofchange.client.AssignCaseAccessClient;
+import uk.gov.hmcts.divorce.solicitor.client.organisation.FindUsersByOrganisationResponse;
+import uk.gov.hmcts.divorce.solicitor.client.organisation.OrganisationClient;
+import uk.gov.hmcts.divorce.solicitor.client.organisation.ProfessionalUser;
+import uk.gov.hmcts.reform.authorisation.generators.AuthTokenGenerator;
+import uk.gov.hmcts.reform.ccd.client.model.AboutToStartOrSubmitCallbackResponse;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+import static uk.gov.hmcts.divorce.divorcecase.model.State.POST_SUBMISSION_STATES;
+import static uk.gov.hmcts.divorce.divorcecase.model.UserRole.APPLICANT_1_SOLICITOR;
+import static uk.gov.hmcts.divorce.divorcecase.model.UserRole.CASE_WORKER;
+import static uk.gov.hmcts.divorce.divorcecase.model.UserRole.JUDGE;
+import static uk.gov.hmcts.divorce.divorcecase.model.UserRole.LEGAL_ADVISOR;
+import static uk.gov.hmcts.divorce.divorcecase.model.UserRole.NOC_APPROVER;
+import static uk.gov.hmcts.divorce.divorcecase.model.UserRole.SUPER_USER;
+import static uk.gov.hmcts.divorce.divorcecase.model.access.Permissions.CREATE_READ_UPDATE;
+import static uk.gov.hmcts.divorce.noticeofchange.model.AcaRequest.acaRequest;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class SystemApplyNoticeOfChange implements CCDConfig<CaseData, State, UserRole> {
+
+    private final  AuthTokenGenerator authTokenGenerator;
+    private final  ObjectMapper objectMapper;
+    private final  AssignCaseAccessClient assignCaseAccessClient;
+    private final  IdamService idamService;
+    private final OrganisationClient organisationClient;
+
+    public static final String NOTICE_OF_CHANGE_APPLIED = "notice-of-change-applied";
+
+    @Override
+    public void configure(final ConfigBuilder<CaseData, State, UserRole> configBuilder) {
+        new PageBuilder(configBuilder
+            .event(NOTICE_OF_CHANGE_APPLIED)
+            .forStates(POST_SUBMISSION_STATES)
+            .name("Notice Of Change Applied")
+            .grant(CREATE_READ_UPDATE, NOC_APPROVER)
+            .grantHistoryOnly(LEGAL_ADVISOR, JUDGE, CASE_WORKER, SUPER_USER)
+            .aboutToStartCallback(this::aboutToStart));
+    }
+
+    public AboutToStartOrSubmitResponse<CaseData, State> aboutToStart(final CaseDetails<CaseData, State> details) {
+        log.info("Applying notice of change for case id: {}", details.getId());
+
+        String sysUserToken = idamService.retrieveSystemUpdateUserDetails().getAuthToken();
+
+        String s2sToken = authTokenGenerator.generate();
+
+        AboutToStartOrSubmitCallbackResponse response =
+            assignCaseAccessClient.applyNoticeOfChange(sysUserToken, s2sToken, acaRequest(details));
+
+        CaseData responseData = objectMapper.convertValue(response.getData(), CaseData.class);
+
+        updateChangeOfRepresentation(details, responseData, sysUserToken, s2sToken);
```

Comment:
Do you think it's best to do this before or after the call to ACA? I'm thinking about whether it's possible that ACA could reassign roles and then some error in this code could result in partially complete NoCs, in which a sol has access to the case but their details have not been saved to it

Reply:
Refactored code as Adam suggested